analyzere/utils.py

```diff
@@ -92,3 +92,18 @@ def vectorize(values):
     if isinstance(values, list):
         return ','.join(str(v) for v in values)
     return values
+
+
+def vectorize_range(values):
+    """
+    This function is for url encoding.
+    Takes a value or a tuple or list of tuples and returns a single result,
+    tuples are joined by "," if necessary, elements in tuple are joined by '_'
+    """
+    if isinstance(values, tuple):
+        return '_'.join(str(i) for i in values)
+    elif isinstance(values, list):
+        if all([isinstance(item, tuple) for item in values]):
+            return ','.join('_'.join(str(i) for i in v) for v in values)
+        raise TypeError('Items in the list must be tuples')
```

Comment:
This is really a style thing so feel free to ignore, but I find [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) are preferable to nested logic.

so the nested `if` would be

```
if not all([isinstance(item, tuple) for item in values]):
    raise TypeError('Items in the list must be tuples')
return ','.join('_'.join(str(i) for i in v) for v in values)
```

It can make the error condition a little easier to read and then lets you proceed knowing certain facts about the data without nesting. 

For the same reason, the `elif` can be just an `if`, since the first `if` is an early exit condition, but that has less much tangible benefit.

I didn't test the above, just eyeballed it.

Reply:
I think this didn't catch the case when the value is a single non-iterable value (like string, int).
I will change `elif -> if`, like:
``` 
if isinstance(values, tuple):
    return '_'.join(str(i) for i in values)
if isinstance(values, list):
    if all([isinstance(item, tuple) for item in values]):
        return ','.join('_'.join(str(i) for i in v) for v in values)
    raise TypeError('Items in the list must be tuples')
return str(values)
```
do you have any ideas to simplfy this?

Reply:
@Fishflavored Do you mean this didn't work?
```
if isinstance(values, tuple):
    return '_'.join(str(i) for i in values)
if isinstance(values, list):
    if not all([isinstance(item, tuple) for item in values]):
        raise TypeError('Items in the list must be tuples')
    return ','.join('_'.join(str(i) for i in v) for v in values)
return str(values)
```

Reply:
Sorry, maybe it wasn't clear in my first comment, I was basically saying I would swap `if all` statement for an `if not all`... just simply because if the logic does get more complex, it is a little easier to follow.

It's not a huge deal here, but hopefully the link illustrates how it can make things a little nicer.

Reply:
oh, I see, that works, sorry I didn't catch `the nested if` in your first comment, good to know that, thanks